### PR TITLE
Handle errors on `generateTeeSession`

### PR DESCRIPTION
**Pull request by mcornaton**

### CODE REVIEW

File: src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java [hunk 1]
```diff
@@ -314,26 +348,38 @@ Map<String, String> getPostComputePalaemonTokens(PalaemonSessionRequest request)
         // encryption
         Map<String, String> encryptionTokens = getPostComputeEncryptionTokens(request);
         if (encryptionTokens.isEmpty()) {
-            throw new Exception("Failed to get post-compute encryption tokens - taskId: " + taskId);
+            // This branch should never happen
+            throw new TeeSessionGenerationException(
+                    POST_COMPUTE_GET_ENCRYPTION_TOKENS_FAILED_UNKNOWN_ISSUE,
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java:353]: I feel we can rename remove `_UNKNOWN_ISSUE` suffix in these

File: src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java [hunk 2]
```diff
@@ -314,26 +348,38 @@ Map<String, String> getPostComputePalaemonTokens(PalaemonSessionRequest request)
         // encryption
         Map<String, String> encryptionTokens = getPostComputeEncryptionTokens(request);
         if (encryptionTokens.isEmpty()) {
-            throw new Exception("Failed to get post-compute encryption tokens - taskId: " + taskId);
+            // This branch should never happen
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java:351]: Why :) ?

Comment (mcornaton) [reply src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java:351]: According to `getPostComputeEncryptionTokens`, `encryptionTokens` should always have at least `RESULT_ENCRYPTION` and `RESULT_ENCRYPTION_PUBLIC_KEY` keys. It's more like a safeguard, if `getPostComputeEncryptionTokens` is ever modified.

File: src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java (hunk 1 above)
Comment (mcornaton) [reply src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java:353]: I like the idea of explicitly telling something went wrong but we do not exactly know what went wrong. That's where `_UNKNOWN_ISSUE` seems useful. Would you really prefer `POST_COMPUTE_GET_ENCRYPTION_TOKENS_FAILED`?

**Review by jbern0rd — APPROVED** [review]

File: src/main/java/com/iexec/sms/tee/session/TeeSessionService.java [hunk 3]
```diff
@@ -78,7 +83,13 @@ public String generateTeeSession(
                 .generateSecureSession(sessionYmlAsString.getBytes())
                 .getStatusCode()
                 .is2xxSuccessful();
-        return isSessionGenerated ? sessionId : "";
+        if (!isSessionGenerated) {
+            throw new TeeSessionGenerationException(
+                    SECURE_SESSION_GENERATION_FAILED,
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/tee/session/TeeSessionService.java:88]:
```suggestion
                    SECURE_SESSION_POST_FAILED,
```
?

File: src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java (hunk 1 above)
Comment (jeremyjams) [reply src/main/java/com/iexec/sms/tee/session/palaemon/PalaemonSessionService.java:353]: I see ok :)

File: src/main/java/com/iexec/sms/tee/TeeController.java [hunk 4]
```diff
@@ -95,16 +111,35 @@ public ResponseEntity<String> generateTeeChallenge(@PathVariable String chainTas
...
+    public ResponseEntity<ApiResponseBody<String, TeeSessionGenerationError>> generateTeeSession(
             @RequestHeader("Authorization") String authorization,
             @RequestBody WorkerpoolAuthorization workerpoolAuthorization) {
         String workerAddress = workerpoolAuthorization.getWorkerWallet();
         String challenge = authorizationService.getChallengeForWorker(workerpoolAuthorization);
         if (!authorizationService.isSignedByHimself(challenge, authorization, workerAddress)) {
-            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+            final ApiResponseBody<String, TeeSessionGenerationError> body =
+                    ApiResponseBody.<String, TeeSessionGenerationError>builder()
+                            .error(REQUEST_NOT_SIGNED_BY_HIMSELF)
```
Comment (jeremyjams) [inline src/main/java/com/iexec/sms/tee/TeeController.java:122]:
```suggestion
                            .error(INVALID_AUTHORIZATION)
```
?

File: iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    EXECUTION_NOT_AUTHORIZED_NO_MATCH_ONCHAIN_TYPE,
+    EXECUTION_NOT_AUTHORIZED_GET_CHAIN_TASK_FAILED,
+    EXECUTION_NOT_AUTHORIZED_TASK_NOT_ACTIVE,
+    EXECUTION_NOT_AUTHORIZED_GET_CHAIN_DEAL_FAILED,
+    EXECUTION_NOT_AUTHORIZED_INVALID_SIGNATURE,
+
+    GET_TASK_DESCRIPTION_FAILED,
+
+    NO_SESSION_REQUEST,
+    NO_TASK_DESCRIPTION,
```
Comment (jeremyjams) [inline iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java:32]:
```suggestion
     // generic errors
    NO_TASK_DESCRIPTION,
    ...
```

File: src/main/java/com/iexec/sms/tee/session/TeeSessionService.java (hunk 3 above)
Comment (mcornaton) [reply src/main/java/com/iexec/sms/tee/session/TeeSessionService.java:88]: What about `SECURE_SESSION_CAS_FAILED`? I find both of our propositions too specific. But maybe users don't want to bother with CAS?

File: iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java (hunk 5 above)
Comment (mcornaton) [reply iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java:32]: What do you mean with `generic errors`? Is that generic opposed to stage-specific values? In that case, there are other values to put under this label.

File: src/main/java/com/iexec/sms/tee/session/TeeSessionService.java (hunk 3 above)
Comment (jeremyjams) [reply src/main/java/com/iexec/sms/tee/session/TeeSessionService.java:88]: I like the idea of adding CAS explicitly yes, there is just a verb missing in your proposition

File: iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java (hunk 5 above)
Comment (jeremyjams) [reply iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java:32]: Yes, other values under this label ( `...`)
If I'm not wrong, some values are thrown in different locations, some are not

Comment (mcornaton) [reply iexec-sms-library/src/main/java/com/iexec/sms/api/TeeSessionGenerationError.java]: The only one that's thrown in different locations is `NO_TASK_DESCRIPTION`.
Nonetheless, I have sorted all the enum values in https://github.com/iExecBlockchainComputing/iexec-sms/pull/93/commits/938bb042fe65a174dab9a85d23254b1d47a1a53b, let me know if this is OK for you :)

**Pull request merged**